lib/pbench/server/database/models/users.py

```diff
@@ -0,0 +1,124 @@
+import datetime
+from flask_bcrypt import generate_password_hash
+from email_validator import validate_email
+from sqlalchemy import Column, Integer, String, DateTime, LargeBinary
+from pbench.server.database.database import Database
+from sqlalchemy.orm import relationship, validates
+
+
+class User(Database.Base):
+    """ User Model for storing user related details """
+
+    __tablename__ = "users"
+
+    id = Column(Integer, primary_key=True, autoincrement=True)
+    username = Column(String(255), unique=True, nullable=False)
+    first_name = Column(String(255), unique=False, nullable=False)
+    last_name = Column(String(255), unique=False, nullable=False)
+    password = Column(LargeBinary(128), nullable=False)
+    registered_on = Column(DateTime, nullable=False, default=datetime.datetime.now())
+    email = Column(String(255), unique=True, nullable=False)
+    auth_tokens = relationship("ActiveTokens", backref="users")
+
+    def __str__(self):
+        return f"User, id: {self.id}, username: {self.username}"
+
+    def get_json(self):
+        return {
+            "username": self.username,
+            "email": self.email,
+            "first_name": self.first_name,
+            "last_name": self.last_name,
+            "registered_on": self.registered_on,
+        }
+
+    @staticmethod
+    def get_protected():
+        return ["registered_on", "id"]
+
+    @staticmethod
+    def query(id=None, username=None, email=None):
+        # Currently we would only query with single argument. Argument need to be either username/id/email
+        if username:
+            user = Database.db_session.query(User).filter_by(username=username).first()
+        elif id:
+            user = Database.db_session.query(User).filter_by(id=id).first()
+        elif email:
+            user = Database.db_session.query(User).filter_by(email=email).first()
+        else:
+            user = None
+
+        return user
+
+    def add(self):
+        """
+        Add the current user object to the database
+        """
+        try:
+            Database.db_session.add(self)
+            Database.db_session.commit()
+        except Exception as e:
+            Database.db_session.rollback()
+            raise e
+
+    @validates("password")
+    def evaluate_password(self, key, value):
+        return generate_password_hash(value)
+
+    # validate the email field
+    @validates("email")
+    def evaluate_email(self, key, value):
+        valid = validate_email(value)
+        email = valid.email
+
+        return email
+
+    def update(self, **kwargs):
+        """
+        Update the current user object with given keyword arguments
+        """
+        try:
+            for key, value in kwargs.items():
+                if key == "auth_tokens":
+                    # Insert the auth token
+                    self.auth_tokens.append(value)
```

Comment:
I believe that SQLAlchemy will remove them from the relationship list when it deletes the item. But I was thinking about this comment when I finished my latest #2074 commit, and I added asserts in the unit test to prove that the `Dataset` metadatas list is empty after a Metadata removal. More SQL alchemy...

Reply:
When the user is created the `auth_tokens` column is an empty list (created by sqlalchemy) and when the user first time logs in we create an auth token for the user and append it to the list. Inside the `ActiveTokens` model we define `user_id` as column that references user as a `ForeignKey` with `ondelete="CASCADE"` so when a user gets deleted it, all the tokens associated with that user also gets deleted.

Reply:
I want to believe that the column is initially empty.  I'm sure that it would be if we fetched the record from the database.  But, the first thing we do is a Python instantiation...and I can't say for sure that sqlalchemy intercedes and prevents Python's rules from applying.  That is, absent sqlalchemy, the class attributes would be used to initialize the instance attributes -- normally, this wouldn't be a problem because, for most fields, we overwrite the value with the one that we want to store -- but, for tokens, we _append_ to the existing value, which we initially hope is empty.  Fortunately, it doesn't really matter...it's just annoying, that's all.

And, I concur that when the user gets deleted, all his/her tokens will be deleted, too.

However, it seems like a user could be long-lived.  It seems like s/he could log in (and log out) hundreds or thousands of times (or, in the case of non-human users, millions of times).  I don't see a mechanism which prevents a huge build-up of old (presumably expired) tokens in the user's database entry.

That is, we have two tables -- users and active tokens -- right?  The tokens table is carefully managed.  However, we are also adding the tokens to the rows in the user table, right?  When/how do we remove tokens from there?

Reply:
@webbnh we dont have to explicitly delete tokens from the user's `auth_tokens` list, it will be automatically removed from the list when the auth token gets deleted from `active_tokens` table because we have established a bidirectional relationship between users and active_tokens table using `backref` keyword argument.
Take a look at this sqlalchemy documentation https://docs.sqlalchemy.org/en/14/orm/backref.html

Reply:
Thanks, Nikhil.  So, SQLAlchemy is truly magical.  :-D  (Actually, I guess it's Python which is magical -- SQLAlchemy is simply wielding that magic.  ;-) )

Anyway, I'm relieved that the class attribute values aren't leaking into the instance attribute values, and I'm glad that we're not going to accumulate piles of expired tokens!